nni/compression/pytorch/speedup/compress_modules.py

```diff
@@ -3,196 +3,364 @@
+    Parameters
+    ---------
+    t_mask: torch.Tensor
+        The tensor only have 1s and 0s, 0 indicates this value is masked
+        and 1 indicates the corresponding value is not masked.
+    dim: int
+        Try to reduce the mask tensor on this dimension.
+
+    Returns
+    -------
+    indexes: torch.Tensor
+        The indexes of the sparsity that can be structurally removed.
+    remained_indexes: torch.Tensor
+        The indexes of values that need to be remained.
+    """
+    assert isinstance(t_mask, torch.Tensor)
+    shape = list(t_mask.size())
+    n_dims = len(shape)
+    dim_list = list(range(n_dims))
+    # try to reduce the mask from the dim-th dimension
+    dim_list.remove(dim)
+
+    t_merged = torch.sum(t_mask, dim_list)
+    assert t_merged.size(0) == shape[dim]
+    all_pruned = t_merged == 0
+    need_remain = t_merged != 0
+    # return the indexes of the sparsity that can be removed
+    indexes = torch.nonzero(all_pruned, as_tuple=True)[0]
+    remained_indexes = torch.nonzero(need_remain, as_tuple=True)[0]
+    return indexes, remained_indexes
+
+
+def no_replace(module, masks):
     """
     No need to replace
     """
     _logger.debug("no need to replace")
     return module
 
 
-def replace_linear(linear, mask):
+def replace_linear(linear, masks):
     """
+    This function will replace the original linear according to
+    the infered masks. This function support the fine-grained and
+    coarse-grained sparsity. In the fine-grained scenario, this function
+    will remove the whole column/row that happen to be totally covered by
+    the masks.
+
     Parameters
     ----------
     linear : torch.nn.Linear
         The linear module to be replace
-    mask : ModuleMasks
-        The masks of this module
+    masks : Tuple of the input masks, output masks and weight masks
+        Tuple of the masks, for example
+        ([input_m1, input_m2], [output_m], {'weight':weight_m})
 
     Returns
     -------
     torch.nn.Linear
         The new linear module
     """
-    assert isinstance(mask, ModuleMasks)
-    assert mask.input_mask is not None
-    assert mask.output_mask is None
-    assert not mask.param_masks
-    index = mask.input_mask.mask_index[-1]
-    in_features = index.size()[0]
-    _logger.debug("replace linear with new in_features: %d", in_features)
-    new_linear = torch.nn.Linear(in_features=in_features,
-                                 out_features=linear.out_features,
-                                 bias=linear.bias is not None)
-    new_linear.to(linear.weight.device)
-    new_linear.weight.data = torch.index_select(
-        linear.weight.data, -1, index.to(linear.weight.device))
+    in_masks, output_mask, weight_mask = masks
+    assert isinstance(linear, nn.Linear)
+    assert len(in_masks) == 1
+    assert isinstance(output_mask, torch.Tensor)
+    in_mask = in_masks[0]
+    # only need the first batch of the constant
```

Comment:
this can be removed? or move to the right place

Reply:
removed, thanks.